File: BAPC/2017/easter.py



n,b,r = list(map(int, input().split()))


bb = [list(map(int, input().split())) for i in range(b)]
rb = [list(map(int, input().split())) for i in range(r)]


from math import ceil,sqrt
def dist(p,q):
    return (p[0]-q[0])*(p[0]-q[0]) + (p[1]-q[1])*(p[1]-q[1])

def max_flow(C, s, t):
    n = len(C)  # C is the capacity matrix
    F = [[0] * n for i in range(n)]
    path = bfs(C, F, s, t)
    #  print path
    while path != None:
        flow = min(C[u][v] - F[u][v] for u, v in path)
        for u, v in path:
            F[u][v] += flow
            F[v][u] -= flow
        path = bfs(C, F, s, t)
    return sum(F[s][i] for i in range(n))


# find path by using BFS
def bfs(C, F, s, t):
    queue = [s]
    paths = {s: []}
    if s == t:
        return paths[s]
    while queue:
        u = queue.pop(0)
        for v in range(len(C)):
            if (C[u][v] - F[u][v] > 0) and v not in paths:
                paths[v] = paths[u] + [(u, v)]
                if v == t:
                    return paths[v]
                queue.append(v)
    return None

def solve(minDist):
    global bb, rb

    sqDist = minDist*minDist
    caps = [[0]*(len(bb)+2+len(rb)) for i in range(len(bb)+len(rb)+2)]
    ## 0 is s, 1 is t, 2...len(bb)+1 is blue len(bb)+2..len(bb)+len(rb)+2 is red
    for bind,b in enumerate(bb):
        caps[0][2+bind] = 1
        #caps[2+bind][0] = 1
        for rind,r in enumerate(rb):
            euclid = dist(b,r)
            #print(b,r, euclid)
            if euclid <= sqDist:
                caps[2+bind][2+len(bb)+rind] = 1
                #caps[2+len(bb)+rind][2+bind] = 1
    for rind,r in enumerate(rb):
        caps[2+len(bb)+rind][1] = 1
        #caps[1][2+len(bb)+rind] = 1
    # for cap in caps:
    #     print(*cap)

    # G = Graph(caps)
    # res = G.FordFulkerson(0,1)
    res = max_flow(caps, 0,1)
    #print("FOUND", res)
    return res





low = 1
high = 30000

EPS = 1e-8
while abs(high-low) > EPS:
    mid = (low+high)/2
    ### Can we solve this where the minimum distance apart is mid
    #print(low, mid, high)
    res = solve(mid)


    if b+r-res >= n:
        ### Can match using this
        low = mid
    else:
        high = mid
print(round(low, 6))

File: BAPC/2017/easter.cpp
#include <bits/stdc++.h>

using namespace std;


double EPS = 1e-9;
typedef long long LL;
typedef vector<pair<int, int>> vpii;
struct Edge {
  int u, v;
  LL cap, flow;
  Edge() {}
  Edge(int u, int v, LL cap): u(u), v(v), cap(cap), flow(0) {}
};

struct Dinic {
  int N;
  vector<Edge> E;
  vector<vector<int>> g;
  vector<int> d, pt;
  
  Dinic(int N): N(N), E(0), g(N), d(N), pt(N) {}

  void AddEdge(int u, int v, LL cap) {
    if (u != v) {
      E.emplace_back(u, v, cap);
      g[u].emplace_back(E.size() - 1);
      E.emplace_back(v, u, 0);
      g[v].emplace_back(E.size() - 1);
    }
  }

  bool BFS(int S, int T) {
    queue<int> q({S});
    fill(d.begin(), d.end(), N + 1);
    d[S] = 0;
    while(!q.empty()) {
      int u = q.front(); q.pop();
      if (u == T) break;
      for (int k: g[u]) {
        Edge &e = E[k];
        if (e.flow < e.cap && d[e.v] > d[e.u] + 1) {
          d[e.v] = d[e.u] + 1;
          q.emplace(e.v);
        }
      }
    }
    return d[T] != N + 1;
  }

  LL DFS(int u, int T, LL flow = -1) {
    if (u == T || flow == 0) return flow;
    for (int &i = pt[u]; i < g[u].size(); ++i) {
      Edge &e = E[g[u][i]];
      Edge &oe = E[g[u][i]^1];
      if (d[e.v] == d[e.u] + 1) {
        LL amt = e.cap - e.flow;
        if (flow != -1 && amt > flow) amt = flow;
        if (LL pushed = DFS(e.v, T, amt)) {
          e.flow += pushed;
          oe.flow -= pushed;
          return pushed;
        }
      }
    }
    return 0;
  }

  LL MaxFlow(int S, int T) {
    LL total = 0;
    while (BFS(S, T)) {
      fill(pt.begin(), pt.end(), 0);
      while (LL flow = DFS(S, T))
        total += flow;
    }
    return total;
  }
};

double sqdist(pair<int, int> p, pair<int, int> q){
    return (p.first-q.first)*(p.first-q.first)+(p.second-q.second)*(p.second-q.second); 

}


int solve(double minDist, vpii blues, vpii reds){

    int tot = blues.size()+reds.size()+2;

    double sqDist = minDist*minDist;

    Dinic graph = Dinic(tot);

    for(int i=0; i < blues.size(); i++){
        graph.AddEdge(0, i+2, 1);

        for(int j=0; j < reds.size(); j++){
            double pdist = sqdist(blues[i], reds[j]);

            if(sqDist-pdist >= EPS){
                graph.AddEdge(i+2, j+blues.size()+2,1);
            }
        }
    }

    for(int i=0; i < reds.size(); i++){
        graph.AddEdge(i+2+blues.size(), 1,1);
    }

    return graph.MaxFlow(0,1);


}

int main(){
    int n,b,r;
    cin >> n >> b >> r;
    vector<pair<int,int>> blues(b);
    vector<pair<int, int>> reds(r);

    for(int i=0; i < b; i++){
        cin >> blues[i].first >> blues[i].second;
    }

    for(int i=0; i < r; i++){
        cin >> reds[i].first >> reds[i].second;
    }

    //cout << solve(1, blues, reds) << "\n";

    double low = 1;
    double high = 30000;


    while(abs(high-low) > EPS){
        double mid = (low+high)/2;
        int res = solve(mid, blues, reds);

        if(r+b-res >= n){
            low = mid;
        }else{
            high = mid;
        }
        
    }
    cout << fixed << setprecision(8) << low << "\n";

    return 0;

}